core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -179,55 +318,73 @@ private[zio] final class FiberContext[E, A](platform: Platform, startEnv: AnyRef
                   curIo = nextInstr(value)
 
                 case ZIO.Tags.EffectTotal =>
-                  val io = curIo.asInstanceOf[ZIO.EffectTotal[Any]]
+                  val io     = curIo.asInstanceOf[ZIO.EffectTotal[Any]]
+                  val effect = io.effect
 
-                  curIo = nextInstr(io.effect())
+                  if (traceEffects && inTracingRegion) addTrace(effect)
```

Comment:
All the testing for `inTracingRegion` can be accelerated (right now it goes through heap access every time).

Reply:
Will do. Although, I have to recheck, but I'm fairly sure it gets JIT'ed away on current benchmarks due to tracingStatus never changing, will need a benchmark that toggles tracing

Reply:
After testing: Putting the StackBool on stack hurts performance in the common case when tracing status is not changed. Using a `var currentStatus: Boolean` instead of the heap also hurts perf (4800 vs. 4900 ops). Seems like using class fields is more jit friendly in this case.

Reply:
Using a `var` means we have to create a `BooleanRef` to be able to change it in `bracket_` which hurts performance and JIT-ability... Adding more primops to toggle status might alleviate that, but I guess it's too much since we're probably getting the best performance for the commont case of non-changing tracing status currently anyway.